crates/zksync/inspectors/src/trace.rs

```diff
@@ -316,15 +324,22 @@ impl InspectorExt for TraceCollector {
                 }
             }
 
-            if !overflow && record_trace {
-                context.journaled_state.depth = context.journaled_state.depth.saturating_sub(1);
-            }
-
             extra_gas
         }
 
+        let (new_depth, overflow) = context.journaled_state.depth.overflowing_add(1);
+        // If we are going to record the top call then we don't want to change the call depth
+        if !overflow && !record_top_call {
+            context.journaled_state.depth = new_depth;
+        }
+
         for call in call_traces {
-            trace_call_recursive(&mut self.inner, context, call, false);
+            trace_call_recursive(&mut self.inner, context, call, record_top_call);
+        }
+
+        // We then decrement the call depth so `call_end`/`create_end` has the correct context
```

Comment:
I didn't put the comment here because we don't have further calls to `self.inner`... this section is more of a cleanup